### PR TITLE
Bump kubernetes-client-bom from 6.3 to 6.4

### LINKED ISSUE #29520: decouple from okhttp(3)
### Description

Currently quarkus ends up having a dependency to okhttp3 via the kubernetes client. This issue tries to gather the issues/options as fixing this potentially involves collaboration across Quarkus, fabric8 io kubernetes client, operator SDK and the OpenJDK team. 

This is a root cause of a list of issues:
1) CVE issue list for okhttp3 keeps growing (albeit at this time of writing not aware of issues that gets exposed via Quarkus usage of okhttp3 but it is problematic if anyone do need to use okhttp for other means)
2) We prefer to use one HTTP client (vert.x) to reduce exposure of CVE's and in vert.x case to be able to reuse the same asyn event loop wherever possible
3) Upgrading to okhttp4 would help on both #1 and #2 but then we drag in a reliance on kotlin stdlib which has all kinds of potential downsides when it comes to dependency alignment and support in native. 

Thus the best option will be to find ways to not introduce more CVE exposure, not add more libraries and still have good async and native support.

On Kubernetes client side they now have support for different HTTP clients. We should make that work for us!

There are 3 client options:

## Jetty 

This client is somewhat complete but has some issues around blocking behaviours. Adding Jetty into Quarkus managed dependency set does not sound like a good option.

Users can though update the dependency to use it if they so wish.

## JDK http client

This is working but only recommended for Java 16 and upwards due to [JDK 11 bug](https://bugs.openjdk.org/browse/JDK-8245245) affecting query parameters in websocket based calls. 

Also would be improved when [cleaner async support](https://github.com/fabric8io/kubernetes-client/pull/4430) merged

The JDK bug we can probably get backported to JDK 11. Will take a few months before broadly available though (January earliest).

## vert.x 

Current progress is in [this branch](https://github.com/fabric8io/kubernetes-client/tree/vertx-client/httpclient-vertx).
It has issues and is awaiting https://github.com/fabric8io/kubernetes-client/pull/4430 to get merged. This PR fixes issues 
around how the client deals with bytebuffers and streams. With this in a vert.x client should be much more manageable to get in.

### Implementation ideas

My (Max) preferable option is to get vert.x client enabled as that avoids all the issues know for okhttp.
Secondary option is to get jdk issue fixed allowing users to at least just fall back to the JDK.

Jetty I'm not a fan of as it adds yet another dependency chain to maintain.

### LINKED ISSUE #30558: Kubernetes extension build error
### Describe the bug

When using the kubernetes extension, there is an error at build time: `Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.`

We've been running fine even with the error, so it is probably harmless. But it is still concerning to see an error in the build.

### Expected behavior

No error

### Actual behavior

Build error:

```
Error reading service account token from: [/var/run/secrets/kubernetes.io/serviceaccount/token]. Ignoring.
```

### How to Reproduce?

Reproducer:

1. Clone https://github.com/mikethecalamity/quarkus-test
2. `./gradlew -p apps service3:build`

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

17

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.15.3

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Gradle 7.5.1

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by manusa**
Kubernetes Client ~6.4.0~ 6.4.1 was just released: https://github.com/fabric8io/kubernetes-client/releases/tag/v6.4.1

This release of the client includes the new Vert.x `HttpClient` implementation. This PR includes changes in the `pom.xml` files to make use of this client instead of the default `OkHttp` implementation. If everything goes well, this should fix #29520

- Added an enforcer rule to ensure ~~OkHttp~~ that the `kubernetes-httpclient-okhttp` module is not mistakenly propagated, however, I'm not sure this is the right place. I can't directly block OkHttp since it is also used (server-side) in the MockServer.
- Modified the kubernetes-extension (creates kubernetes et al manifests and deploys them) to create and close KubernetesClient instances and avoid the leakages mentioned in some of the PR comments
- Modified the kubernetes-client-extension to implement an HttpClient.Factory that reuses the global Vert.x instance (uses CDI to retrieve it, maybe doesn't work, waiting for tests)

/cc @metacosm @maxandersen @shawkins @gastaldi @geoand 

Fix https://github.com/quarkusio/quarkus/issues/30558

### CODE REVIEW

**Review by gastaldi — APPROVED** [review]
That's awesome! You can also add the exclusion to the `bom/application` pom.xml

**Comment by gastaldi** [comment]
> That's awesome! You can also add the exclusion to the `bom/application` pom.xml

Ah but we only import the kubernetes-bom there, nevermind

**Review by gsmet — APPROVED** [review]
Awesome!

As a followup, I will try to remove okhttp from the BOM as it shouldn't be a dependency anymore.

**Review by gastaldi — DISMISSED** [review]
I think it's safe to remove OkHttp from the list of managed dependencies too:
 https://github.com/quarkusio/quarkus/blob/main/bom/application/pom.xml#L171-L172

and 

https://github.com/quarkusio/quarkus/blob/main/bom/application/pom.xml#L4169-L4183

**Comment by gsmet** [comment]
Yeah I think it's better to do it in a separate PR in case it breaks something and we have to revert it.

**Review by geoand — APPROVED** [review]
Very glad to see this!

**Comment by gsmet** [comment]
An example of the enforcer failures:
```
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce) @ quarkus-kubernetes-client-internal ---
[18537](https://github.com/quarkusio/quarkus/actions/runs/3959243390/jobs/6781818473#step:10:18538)
Warning:  Rule 3: org.apache.maven.plugins.enforcer.ExternalRules failed with message:
[18538](https://github.com/quarkusio/quarkus/actions/runs/3959243390/jobs/6781818473#step:10:18539)
Found Banned Dependency: com.squareup.okhttp3:okhttp:jar:3.14.9
[18539](https://github.com/quarkusio/quarkus/actions/runs/3959243390/jobs/6781818473#step:10:18540)
Use 'mvn dependency:tree' to locate the source of the banned dependencies.
```

**Comment by manusa** [comment]
It seems that we might need to bump Dekorate along with the client.

https://github.com/dekorateio/dekorate/pull/1131

**Comment by geoand** [comment]
Including @Sgitario for awareness

**Comment by Sgitario** [comment]
> Including @Sgitario for awareness

PR to bump dekorate: https://github.com/quarkusio/quarkus/pull/30533

**Comment by manusa** [comment]
Log excerpt for failure:
```
2023-01-24T21:56:54.9279414Z [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 5.406 s <<< FAILURE! - in io.quarkus.it.kubernetes.BasicOpenshiftTest
2023-01-24T21:56:54.9292273Z [ERROR] io.quarkus.it.kubernetes.BasicOpenshiftTest  Time elapsed: 5.406 s  <<< ERROR!
2023-01-24T21:56:54.9292894Z java.lang.RuntimeException: 
2023-01-24T21:56:54.9293556Z java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
2023-01-24T21:56:54.9294898Z 	[error]: Build step io.quarkus.kubernetes.client.deployment.KubernetesClientBuildStep#process threw an exception: java.util.ServiceConfigurationError: io.fabric8.kubernetes.client.http.HttpClient$Factory: Provider io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory not found
2023-01-24T21:56:54.9295775Z 	at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:589)
2023-01-24T21:56:54.9296281Z 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.nextProviderClass(ServiceLoader.java:1212)
2023-01-24T21:56:54.9296846Z 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1221)
2023-01-24T21:56:54.9297378Z 	at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265)
2023-01-24T21:56:54.9298039Z 	at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300)
2023-01-24T21:56:54.9298438Z 	at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385)
2023-01-24T21:56:54.9299010Z 	at io.fabric8.kubernetes.client.utils.HttpClientUtils.getFactory(HttpClientUtils.java:163)
2023-01-24T21:56:54.9299602Z 	at io.fabric8.kubernetes.client.utils.HttpClientUtils.getHttpClientFactory(HttpClientUtils.java:149)
2023-01-24T21:56:54.9300449Z 	at io.fabric8.kubernetes.client.utils.HttpClientUtils.createHttpClient(HttpClientUtils.java:144)
2023-01-24T21:56:54.9301019Z 	at io.fabric8.kubernetes.client.DefaultKubernetesClient.<init>(DefaultKubernetesClient.java:56)
2023-01-24T21:56:54.9301747Z 	at io.quarkus.kubernetes.client.runtime.KubernetesClientUtils.createClient(KubernetesClientUtils.java:52)
2023-01-24T21:56:54.9302417Z 	at io.quarkus.kubernetes.client.deployment.KubernetesClientBuildStep.process(KubernetesClientBuildStep.java:16)
```

For some reason the SPI `META-INF/services` file including the `io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory` seems to still be processed.

**Comment by geoand** [comment]
I will have a look

**Comment by geoand** [comment]
@Sgitario @iocanel the problem stems from the fact that io/dekorate/servicebinding-annotations/3.3.0/servicebinding-annotations-3.3.0-noapt.jar contains a `META-INF/services/io.fabric8.kubernetes.client.http.HttpClient$Factory` file (that uses `io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory`)

**Comment by manusa** [comment]
> @Sgitario @iocanel the problem stems from the fact that `io/dekorate/servicebinding-annotations/3.3.0/servicebinding-annotations-3.3.0-noapt.jar` contains a `META-INF/services/io.fabric8.kubernetes.client.http.HttpClient$Factory` file (that uses `io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory`)

:+1: I'm not sure how were you able to trace that, I was getting crazy trying to discover the culprit jar.

I provided yet another commit switching to use an explicit factory instead of relying on SPI to provide it. This might do the job, WDYT?

**Comment by Sgitario** [comment]
> > @Sgitario @iocanel the problem stems from the fact that `io/dekorate/servicebinding-annotations/3.3.0/servicebinding-annotations-3.3.0-noapt.jar` contains a `META-INF/services/io.fabric8.kubernetes.client.http.HttpClient$Factory` file (that uses `io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory`)
> 
> +1 I'm not sure how were you able to trace that, I was getting crazy trying to discover the culprit jar.
> 
> I provided yet another commit switching to use an explicit factory instead of relying on SPI to provide it. This might do the job, WDYT?

I guess the problematic line in Dekorate is [this](https://github.com/dekorateio/dekorate/blob/ee7fa59beda60a6f6fb01c8d56be16b1dee0278b/annotations/servicebinding-annotations/pom.xml#L207). This was done as part of these changes: https://github.com/dekorateio/dekorate/pull/790
Though, I don't fully follow why is required and also I don't see it in all the annotations. Maybe @iocanel can provide some light on this.

**Comment by geoand** [comment]
> I'm not sure how were you able to trace that, I was getting crazy trying to discover the culprit jar.

I basiically did `TCCL.getResources("META-INF/services/....")` (or `findResources`, I don't remember) and went through the `Enumeration` looking at each URL

**Comment by iocanel** [comment]
> I guess the problematic line in Dekorate is [this](https://github.com/dekorateio/dekorate/blob/ee7fa59beda60a6f6fb01c8d56be16b1dee0278b/annotations/servicebinding-annotations/pom.xml#L207). This was done as part of these changes: [dekorateio/dekorate#790](https://github.com/dekorateio/dekorate/pull/790)
> Though, I don't fully follow why is required and also I don't see it in all the annotations. Maybe @iocanel can provide some light on this.

The fabric8 client does not provide support for service binding, so the code needed for serializing / deserializing etc is baked directly into dekorate. This also means that it has fallen out of sync with the client (and unfortunately more than once). Let's apply everything needed in dekorate and then create a proper service binding extension for the client.

**Comment by geoand** [comment]
> I provided yet another commit switching to use an explicit factory instead of relying on SPI to provide it. This might do the job, WDYT?

That would break users who tried to exclude the Vert.x client and wanted to rely on OkHTTP though, wouldn't it?

**Comment by iocanel** [comment]
There is still something that I don't get. In dekorate we only define a servcie for `KubenretesResource` that is not related in any way with http client factories and what not.

**Comment by manusa** [comment]
> > I provided yet another commit switching to use an explicit factory instead of relying on SPI to provide it. This might do the job, WDYT?
> 
> That would break users who tried to exclude the Vert.x client and wanted to rely on OkHTTP though, wouldn't it?

I'm not sure if these methods to create the client are used outside of the build process. But yes, that would prevent users from providing alternative HttpClient implementations.

Shall I remove this commit and wait for a solution in Dekorate then?

**Comment by geoand** [comment]
> Shall I remove this commit and wait for a solution in Dekorate then?

If @iocanel and @Sgitario think we can get a solution quickly, then yes, I would prefer that

**Comment by iocanel** [comment]
Ohhhh now I get it. The shade plugin drags in more things than it needs. That's should be simple to fix.

**Comment by Sgitario** [comment]
> > Shall I remove this commit and wait for a solution in Dekorate then?
> 
> If @iocanel and @Sgitario think we can get a solution quickly, then yes, I would prefer that

Fix https://github.com/dekorateio/dekorate/pull/1137

**Comment by Sgitario** [comment]
Pull request with new Dekorate: https://github.com/quarkusio/quarkus/pull/30627

**Comment by manusa** [comment]
```
[INFO] Running io.quarkus.it.kubernetes.KubernetesWithFlywayInitTest
[ERROR] Java heap space
```
Looking into this

**Comment by geoand** [comment]
There is definitely some memory leak... The question is whether it was introduced by this PR or just exarbated for some reason

**Comment by manusa** [comment]
I'm having a hard time figuring out what's causing the leak.

The heap dump reveals that the QuarkusClassLoader is getting huge:
```
48 instances of io.quarkus.bootstrap.classloading.QuarkusClassLoader, loaded by jdk.internal.loader.ClassLoaders$AppClassLoader @ 0xffe57508 occupy 1,127,281,408 (78.32%) bytes.
```

I delved around the `QuarkusProdModeTest` and the `QuarkusBootstrap`, but those seemed to take care of cleaning up after each test execution.

It would be good if someone smarter and more knowledgeable could also take a look into this.

**Comment by geoand** [comment]
I don't have much time to look into it today, but a quick and dirty check shows that the current `main` does not suffer from a ClassLoader leak:

main

When trying to run the same test however with this PR, things get busted **really** quick.

**Comment by geoand** [comment]
A really quick look points indicated that there might be an issue with Vert'x blocked thead checked, but I'll need to investigate more.

**Review by geoand — CHANGES_REQUESTED** [review]
@manusa I am starting to look into the Vert.x integration of the k8s client, but I immediately have a few observations:

It doesn't look like the Vertx object created by Quarkus is being used. This will be a problem for us - we really need to have a single Vertx object to have the best possible runtime footprint (among other things).

So we will either need:

* Our own Quarkus specific `VertxHttpClientFactory`
* Or a way to configure the client's provided `VertxHttpClientFactory` (perhaps using another SPI)

**Comment by geoand** [comment]
There is a chance the memory leak has to do with the `BlockedThreadChecker`. 
@cescoffier I don't see a way to turn that off completely, or am I missing something?

**Comment by cescoffier** [comment]
you can't disable the thread checker altogether. You can increase the threshold. 

However, in that case, I may suggest making sure that the Vert.x HTTP clients are correctly closed after each usage. Also, I'm a bit surprised not to see it using the managed vert.x instance. Without that, it would:

- require to have that vert.x instance closes every time - that's probably the leak
- have a completely different configuration for that vert.x instance (disable async DNS at least)

**Comment by geoand** [comment]
The problem seems to be that because the Fabric8 K8s client does **not** use our Vertx, the object is not getting closed, which means it ends up leaking a ton of stuff

**Comment by geoand** [comment]
So as I mentioned before (and as Clement also implies), we really need to have the client use our Vert.x instance.

**Comment by geoand** [comment]
> So as I mentioned before (and as Clement also implies), we really need to have the client use **our** Vert.x instance - anything else will be very problematic.

@manusa this is an action item for you :). Let me know if I can help

**Comment by metacosm** [comment]
Well, technically, 6.4.0 update could be merged without the switch to the Vert.x client… but it would indeed be better with it 😉

**Comment by geoand** [comment]
I'll leave that up to you guys, but my personal _feeling_ is that it would be better for both projects if the version bump included the working Vert.x integration.

**Comment by manusa** [comment]
> > So as I mentioned before (and as Clement also implies), we really need to have the client use **our** Vert.x instance - anything else will be very problematic.
> 
> @manusa this is an action item for you :). Let me know if I can help

Thx for the analysis. I'm glad I could create a leak big enough to prove the `QuarkusProdTest` useful :sweat_smile: 

I'll give it a try and see if I can create a Quarkus-specific `HttpClient` factory to reuse the Vert.x object. I'll probably ping you with a few questions once I get my hands dirty.

**Comment by geoand** [comment]
> I'll give it a try and see if I can create a Quarkus-specific HttpClient factory to reuse the Vert.x object. I'll probably ping you with a few questions once I get my hands dirty.

👍🏼

**Comment by geoand** [comment]
Before I forget: We will probably also need a way to configure the `HttpClientOptions` being used 😉

**Comment by geoand** [comment]
If this passes CI, I want to also have a closer look under debug

**Comment by geoand** [comment]
I tested this and it works very nicely.

Great job!

**Comment by geoand** [comment]
The one ask I have is to rebase on `main` as this is a little behind now

**Comment by manusa** [comment]
> The one ask I have is to rebase on `main` as this is a little behind now

Yes, I also need to merge your changes to the Knative deployer.


Waiting to the CInin case there is anything else that needs to be addressed

**Comment by manusa** [comment]
Native failure:
```
2023-01-31T17:20:51.7230714Z Caused by: java.lang.IllegalStateException: Unable to locate CDIProvider
2023-01-31T17:20:51.7239722Z 	at javax.enterprise.inject.spi.CDI.findAllProviders(CDI.java:121)
2023-01-31T17:20:51.7240540Z 	at javax.enterprise.inject.spi.CDI.getCDIProvider(CDI.java:82)
2023-01-31T17:20:51.7244500Z 	at javax.enterprise.inject.spi.CDI.current(CDI.java:64)
2023-01-31T17:20:51.7245274Z 	at io.quarkus.kubernetes.client.runtime.QuarkusHttpClientFactory.newBuilder(QuarkusHttpClientFactory.java:18)
2023-01-31T17:20:51.7246208Z 	at io.quarkus.kubernetes.client.runtime.QuarkusHttpClientFactory.newBuilder(QuarkusHttpClientFactory.java:10)
2023-01-31T17:20:51.7250102Z 	at io.fabric8.kubernetes.client.http.HttpClient$Factory.newBuilder(HttpClient.java:44)
2023-01-31T17:20:51.7251116Z 	at io.quarkus.kubernetes.client.runtime.QuarkusHttpClientFactory.newBuilder(QuarkusHttpClientFactory.java:13)
2023-01-31T17:20:51.7260232Z 	at io.fabric8.kubernetes.client.KubernetesClientBuilder.getHttpClient(KubernetesClientBuilder.java:88)
2023-01-31T17:20:51.7271532Z 	at io.fabric8.kubernetes.client.KubernetesClientBuilder.build(KubernetesClientBuilder.java:78)
2023-01-31T17:20:51.7272392Z 	at io.fabric8.kubernetes.client.server.mock.KubernetesMockServer.createClient(KubernetesMockServer.java:128)
2023-01-31T17:20:51.7281657Z 	at io.fabric8.kubernetes.client.server.mock.KubernetesMockServer.createClient(KubernetesMockServer.java:123)
2023-01-31T17:20:51.7285441Z 	at io.quarkus.test.kubernetes.client.KubernetesMockServerTestResource.getClient(KubernetesMockServerTestResource.java:17)
2023-01-31T17:20:51.7295700Z 	at io.quarkus.test.kubernetes.client.KubernetesMockServerTestResource.getClient(KubernetesMockServerTestResource.java:11)
2023-01-31T17:20:51.7300075Z 	at io.quarkus.test.kubernetes.client.AbstractKubernetesTestResource.start(AbstractKubernetesTestResource.java:28)
2023-01-31T17:20:51.7300879Z 	at io.quarkus.test.common.TestResourceManager$TestResourceEntryRunnable.run(TestResourceManager.java:481)
2023-01-31T17:20:51.7301376Z 	... 4 more
``` 

Related to the following snippet:
```java
    @Override
    public VertxHttpClientBuilder<QuarkusHttpClientFactory> newBuilder() {
        return new VertxHttpClientBuilder<>(this, CDI.current().select(Vertx.class).get());
    }
```

Maybe doing the wrong thing here. Any suggestions? try-catch and use `Vertx.vertx()` in case of exception?

**Comment by geoand** [comment]
This happens because the mock server is launched before Quarkus.

We'll need to fallback to creating Vertx if CDI is not available

**Comment by manusa** [comment]
Ouch, didn't think about that one. Fixing.

**Comment by gsmet** [comment]
Don't worry about the Gradle tests failing, it's unrelated to your patch.

**Comment by manusa** [comment]
> Don't worry about the Gradle tests failing, it's unrelated to your patch.

Thx. Yes, I discussed internally with Georgios. I've rebased after the merge of #30758.

Hoping this time we get a completely green report :crossed_fingers: .

**Review by geoand — APPROVED** [review]

**Pull request merged**
